PYIC-8647: handle unexpected error event for driving licence, passport, exper…

thebauSoftwire opened the pull request:
…ian kbv and f2f CRIs

**NOTE: this should not be merged until UCD have reviewed and approved these changes**

## Proposed changes
### What changed

- adding routing for unexpected error event for driving licence, passport, experian kbv and f2f CRIs

### Why did it change

- so that users can be re-directed to try again or retry an alternative identity proofing path if they encounter an unexpected CRI error

### Issue tracking
<!-- Jira ticket & other docs, like RFCs -->

- [PYIC-8647](https://govukverify.atlassian.net/browse/PYIC-8647)

## Checklists

- [ ] READMEs and documentation up-to-date
- [x] API/ unit/ contract tests have been written/ updated
- [x] No risk of exposure: PII, credentials, etc through logs/ code
- [x] Production changes appropriately staged out


[PYIC-8647]: https://govukverify.atlassian.net/browse/PYIC-8647?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

thebauSoftwire committed:
PYIC-8647: handle error event for driving licence, passport and experian kbv CRIs

thebauSoftwire committed:
PYIC-8647: handle error event for f2f CRI

blakeyp commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/f2f-hand-off.yaml:
```diff
@@ -93,11 +99,41 @@ states:
+            targetJourney: NEW_P1_IDENTITY
+            targetState: DCMAW_ASYNC_ANOTHER_WAY
+      returnToRp:
+        targetState: RETURN_TO_RP
```
(same for all other occurrences of `returnToRp`)

blakeyp commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml:
```diff
@@ -47,6 +47,11 @@ states:
+  IDENTITY_DOCUMENT_START:
+    events:
+      next:
+        targetState: IDENTITY_START_PAGE
```
I think we're best going through `RESET_SESSION_IDENTITY` first? we don't really want the existing VCs hanging about

blakeyp commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/f2f-hand-off.yaml:
```diff
@@ -93,11 +99,41 @@ states:
         targetJourney: TECHNICAL_ERROR
         targetState: ERROR_NO_TICF
+        checkJourneyContext:
+          mediumConfidence:
+            targetJourney: NEW_P2_IDENTITY
+            targetState: DCMAW_ASYNC_ANOTHER_WAY
+          lowConfidence:
+            targetJourney: NEW_P1_IDENTITY
+            targetState: DCMAW_ASYNC_ANOTHER_WAY
```
I think we're also gonna want to run through reset-session here too, especially as the combo of CIC + App VCs can be problematic (for name correlation)

arthurs-gds committed:
PYIC-8647 Add reset-session to some routs after error

A reset-session is needed to ensure no name clash happens after
faile F2F

Co-authored-by: Thea Bautista <thea.bautista@digital.cabinet-office.gov.uk>

arthurs-gds committed:
PYIC-8647 add PROCESS_INCOMPLETE_IDENTITY step to error journeys

This it needed for fraud checks before repeating journeys

arthurs-gds committed:
PYIC-8647 refactor routing around SorryTechnicalProblem page

arthurs-gds committed:
PYIC-8647 put sorry-technical-error routing behind a feature flag

thebauSoftwire commented on local-running/core.local.params.yaml:
```diff
@@ -358,6 +358,7 @@ core:
     sqsAsync: true
     kidJarHeaderEnabled: true
     drivingLicenceAuthCheck: true
+    sorryTechnicalErrorEnabled: true
```
should this be set to false by default for now and only enable it for the appropriate tests whilst we develop this? So that the api tests cover both existing behaviour live in prod (to cover regression) and this new feature we're working on

arthurs-gds replied:
Yes, agreed.

arthurs-gds commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/f2f-hand-off.yaml:
```diff
@@ -93,9 +99,67 @@ states:
         targetJourney: FAILED
         targetState: FAILED_NO_TICF
       error:
+        checkFeatureFlag:
```
`CRI_F2F` inherits from `CRI_STATE` which has exactly that default defined. (I had it there but then removed to avoid duplication)

thebauSoftwire replied:
ah so this `error` event will completely override the entire `error` event object defined in `CRI_STATE` so when the State Machine receives an `error` event for this state, it will only see:
```
checkFeatureFlag:
  sorryTechnicalErrorEnabled:
    targetState: ...
```
Sorry, I should have made it clearer in the docs, I'll make that update

arthurs-gds commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/nested-journeys/kbvs.yaml:
```diff
@@ -43,7 +43,23 @@ nestedJourneyStates:
+      error:
+        checkFeatureFlag:
+          sorryTechnicalErrorEnabled:
+            targetState: SORRY_CRI_TECHNICAL_PROBLEM_EXPERIAN_KBV_CRI
```
Also inheriting from `CRI_STATE` here.

arthurs-gds committed:
PYIC-8647 refine and correct api tests to validate sorry-technical flag

The old test coverage didn't include tests for server-errors on the cris
where the new functionality is introduced so a test for flag
disabled is needed. The tests for the new feature also
needed the flag activated in a different way than initailly implemented.
A default state for errors is needed in the yaml because the presence
of the feature flag check overrides the inherited default.

arthurs-gds commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/nested-journeys/web-dl-or-passport.yaml:
```diff
@@ -38,6 +38,23 @@ nestedJourneyStates:
+      error:
+        checkFeatureFlag:
+          sorryTechnicalErrorEnabled:
+            targetState: SORRY_CRI_TECHNICAL_PROBLEM_CRI_UK_PASSPORT
```
As you noted on slack, the feature-flag check overrides the `CRI_STATE` default so it is in fact needed here and in the other cases.

thebauSoftwire commented on api-tests/features/p2-f2f-journey.feature:
```diff
@@ -152,6 +152,19 @@ Feature: P2 F2F journey
+        | Attribute          | Values                                      |
+        | evidence_requested | {"scoringPolicy":"gpg45","strengthScore":3} |
+      Then I get a 'sorry-technical-problem' page response with context 'f2fCriError'
+
```
I don't think we need this test since it's already being tested within the new [unexpected-cri-error.feature](https://github.com/govuk-one-login/ipv-core-back/blob/b64a53e765d3139163ad3ddb23c27e13e1a19708/api-tests/features/unexpected-cri-error.feature#L236-L258) file
